### PR TITLE
ci: add pkgdb lint check

### PR DESCRIPTION
**Pull request by aakropotkin**
Restores GitHub Action from original `pkgdb` repository which runs
`clang-tidy` ( a lint checker ) and converts it output into GitHub suggestions.

This PR also causes existing CI workflows to cancel in progress runs if another
PR for the same _ref_ is pushed.
This should hopefully reduce CI spend.

### CODE REVIEW

File: .github/workflows/lint.yml [hunk 1]
```diff
@@ -0,0 +1,169 @@
+# ============================================================================ #
+#
+# Run `clang-tidy' over the codebase and report warnings as annotations.
+#
+# ---------------------------------------------------------------------------- #
+
+name: 'Lint'
```
Comment (zmitchell) [inline .github/workflows/lint.yml:7]:
```suggestion
name: 'Pkgdb Lint'
```
This isn't running the Rust linter, so we should be explicit about that.

File: .github/workflows/lint.yml [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+      # If the workflow was triggered explicitly then we _might_ be
+      # explicitly asked to check all files using an input param.
+      - name: 'Create Target List ( PR )'
+        if: ${{ github.event.pull_request }}
+        run: |
+          git fetch origin main;
+          echo "Checking changed sources" >&2;
+          ./pkgdb/build-aux/changed-sources.sh > targets;
+
+      - name: 'Create Target List ( Dispatch )'
```
Comment (zmitchell) [inline .github/workflows/lint.yml:47]: Question: What is the `( Dispatch )` supposed to mean here?

**Comment by zmitchell** [comment]
LGTM other than two minor comments

File: .github/workflows/lint.yml (hunk 2 above)
Comment (aakropotkin) [reply .github/workflows/lint.yml:47]: it's a workflow trigger category

File: .github/workflows/lint.yml [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+        uses: 'actions/checkout@v4'
+
+      # Create a list of target files.
+      # If the workflow was triggered by a PR, only check changed files.
+      # If the workflow was triggered explicitly then we _might_ be
+      # explicitly asked to check all files using an input param.
+      - name: 'Create Target List ( PR )'
+        if: ${{ github.event.pull_request }}
+        run: |
+          git fetch origin main;
```
Comment (tomberek) [inline .github/workflows/lint.yml:43]: non-blocking: This check against main. Usually a good assumption, but might confuse if PR'ing onto another target branch.

Comment (aakropotkin) [reply .github/workflows/lint.yml]: I used a variable instead which hopefully makes it a bit more robust

**Review by zmitchell — APPROVED** [review]

**Pull request merged**